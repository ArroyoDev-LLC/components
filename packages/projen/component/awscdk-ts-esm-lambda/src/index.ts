export * from './awscdk-ts-esm-lambda'
