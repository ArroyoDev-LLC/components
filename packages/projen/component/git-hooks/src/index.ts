export * from './git-hooks'
export * from './lint-staged'
