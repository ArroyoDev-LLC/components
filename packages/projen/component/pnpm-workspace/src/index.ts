export * from './pnpm-workspace.ts'
