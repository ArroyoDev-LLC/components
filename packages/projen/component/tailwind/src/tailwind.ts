import {
	TypeScriptSourceFile,
	type TypeScriptSourceFileTransform,
} from '@arroyodev-llc/projen.component.typescript-source-file'
import {
	addPropertyAssignmentsFromObject,
	findComponent,
} from '@arroyodev-llc/utils.projen'
import {
	Component,
	DependencyType,
	type Project,
	type typescript,
} from 'projen'
import type { Config } from 'tailwindcss/types'
import {
	type ImportDeclarationStructure,
	type ObjectLiteralExpression,
	type OptionalKind,
	type SourceFile,
	SyntaxKind,
	type WriterFunction,
} from 'ts-morph'

export interface TailwindOptionsPlugin {
	name: string
	moduleImport: OptionalKind<ImportDeclarationStructure>
	/**
	 * Plugin definition specification.
	 * Defaults to `<name>(<options>)`.
	 */
	spec?: WriterFunction | readonly (string | WriterFunction)[]
}

export interface TailwindOptions {
	config?: Config
}

export class Tailwind extends Component {
	public static of(project: Project): Tailwind | undefined {
		return findComponent(project, Tailwind)
	}

	readonly file: TypeScriptSourceFile

	constructor(
		project: typescript.TypeScriptProject,
		options?: TailwindOptions
	) {
		super(project)
		this.project.deps.addDependency('tailwindcss', DependencyType.RUNTIME)

		this.file = new TypeScriptSourceFile(project, 'tailwind.ts', {
			source: [
				`/** @type {import('tailwindcss').Config} */`,
				'export default {}',
			].join('\n'),
			recreate: true,
		})

		if (options?.config) {
			this.addConfig(options.config)
		}
	}

	addPlugin(plugin: TailwindOptionsPlugin): this {
		const { moduleImport, name, spec } = plugin
		const pluginSpec = spec ?? ((writer) => writer.write(name))
		this.project.deps.addDependency(
			moduleImport.moduleSpecifier,
			DependencyType.RUNTIME
		)
		this.file.addImport(moduleImport)
		this.addConfigTransform((configExpr) => {
			const existsPlugins =
				configExpr
					.getProperty('plugins')
					?.asKind?.(SyntaxKind.PropertyAssignment) ??
				configExpr.addPropertyAssignment({
					name: 'plugins',
					initializer: '[]',
				})

			existsPlugins
				.getInitializerIfKindOrThrow(SyntaxKind.ArrayLiteralExpression)
				.addElements(pluginSpec)
		})
		return this
	}

	addConfig(config: Config): this {
		this.addConfigTransform((configExpr) => {
			addPropertyAssignmentsFromObject(configExpr, config)
		})
		return this
	}

	addConfigTransform(
		transform: (
			configObjectLiteral: ObjectLiteralExpression,
			sourceFile: SourceFile
		) => void
	) {
		const transformer: TypeScriptSourceFileTransform = (src: SourceFile) => {
			const configExport = src
				.getExportAssignments()[0]!
				.getExpressionIfKindOrThrow(SyntaxKind.ObjectLiteralExpression)
			transform(configExport, src)
		}
		this.file.addTransformer(transformer)
		return this
	}
}
