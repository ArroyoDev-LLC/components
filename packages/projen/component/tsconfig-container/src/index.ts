export * from './tsconfig-container.ts'
