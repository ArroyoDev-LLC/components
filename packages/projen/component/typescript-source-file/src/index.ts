export * from './typescript-source-file'
export * from './typescript-source-config'
export * from './type-script-source-file-cache'
