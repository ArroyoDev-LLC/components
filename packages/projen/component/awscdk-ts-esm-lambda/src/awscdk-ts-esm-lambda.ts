import { existsSync } from 'node:fs'
import { TypeScriptSourceFile } from '@arroyodev-llc/projen.component.typescript-source-file'
import { findComponent, isComponent } from '@arroyodev-llc/utils.projen'
import {
	awscdk,
	Component,
	type FileBase,
	type Project,
	TextFile,
	typescript,
	javascript,
} from 'projen'

export interface AwsCdkTsEsmLambdaOptions {
	/**
	 * Project relative path to lambda function construct.
	 */
	readonly filePath: string
}

/**
 * AWS CDK TypeScript Lambda function with ESM support.
 */
export class AwsCdkTsEsmLambda extends Component {
	public static of(project: Project): AwsCdkTsEsmLambda | undefined {
		return findComponent(project, AwsCdkTsEsmLambda)
	}

	readonly lambdaFile: FileBase
	readonly tsFile: TypeScriptSourceFile

	constructor(project: Project, options: AwsCdkTsEsmLambdaOptions) {
		super(project)

		this.lambdaFile = project.tryFindFile(options.filePath) as FileBase

		const lambdaComponents = this.project.components.filter((comp) =>
			isComponent(awscdk.LambdaFunction, comp),
		)
		this.node.addDependency(...lambdaComponents.map((comp) => comp.node))

		const sourceCode = this.project.components.find(
			(comp) =>
				isComponent(TextFile, comp) && comp.path === this.lambdaFile.path,
		) as TextFile
		// @ts-expect-error would required replacing component or other post synth hackarounds otherwise since we need to replace the file before it gets written
		// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment,@typescript-eslint/no-unsafe-call,@typescript-eslint/no-unsafe-member-access
		const source: string = sourceCode.lines.join('\n')
		this.project.tryRemoveFile(this.lambdaFile.path)

		if (!(this.project instanceof typescript.TypeScriptProject)) {
			throw new Error('AwsCdkTsEsmLambda must be used in a TypeScriptProject')
		}

		this.tsFile = new TypeScriptSourceFile(this.project, this.lambdaFile.path, {
			marker: false,
			recreate: true,
			source,
			format: false,
		})
		this.applyTransforms()
		javascript.Eslint.of(project)?.addIgnorePattern?.(this.tsFile.path)
	}

	/**
	 * Apply transforms to the lambda function source file.
	 * @protected
	 */
	protected applyTransforms() {
		this.tsFile.addImport({
			namedImports: ['fileURLToPath'],
			moduleSpecifier: 'node:url',
		})
		this.tsFile.addTransformer((sourceFile) => {
			const functionClass = sourceFile.getClasses()[0]
			if (!functionClass)
				throw new Error('Could not find class in entrypoint file')
			const functionCtor = functionClass.getConstructors()[0]
			if (!functionCtor)
				throw new Error('Could not find class constructor in entrypoint file')
			functionCtor.insertStatements(0, [
				'const __filename = fileURLToPath(import.meta.url);',
				'const __dirname = path.dirname(__filename);',
			])
			sourceFile.formatText()
		})
	}
}

/**
 * Automatically discover and create AWS CDK TypeScript Lambda functions with ESM support.
 */
export class AwsCdkTsEsmLambdaAutoDiscover extends Component {
	constructor(project: Project) {
		super(project)

		const lambdaFunctionFiles = project.files
			.filter((file) => file.path.endsWith('-function.ts'))
			.filter((file) =>
				existsSync(file.absolutePath.replace('-function.ts', '.lambda.ts')),
			)

		for (const lambdaFunctionFile of lambdaFunctionFiles) {
			new AwsCdkTsEsmLambda(project, { filePath: lambdaFunctionFile.path })
		}
	}
}
