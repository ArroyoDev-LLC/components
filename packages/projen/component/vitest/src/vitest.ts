import path from 'node:path'
import { TypeScriptSourceConfig } from '@arroyodev-llc/projen.component.typescript-source-file'
import { Vite } from '@arroyodev-llc/projen.component.vite'
import { cwdRelativePath } from '@arroyodev-llc/utils.projen'
import { Component, JsonFile, type Project, type Task } from 'projen'
import { type TypeScriptProject } from 'projen/lib/typescript'
import { deepMerge } from 'projen/lib/util'
import { type UserWorkspaceConfig } from 'vitest/config'

export enum VitestConfigType {
	WORKSPACE = 'Workspace',
	PROJECT = 'Project',
	CONFIG = 'Config',
}

export interface VitestOptions {
	/**
	 * Vitest config path.
	 */
	configFilePath?: string
	/**
	 * Configuration type.
	 */
	configType?: VitestConfigType
	/**
	 * Vitest settings.
	 */
	settings?: UserWorkspaceConfig
}

export class Vitest extends Component {
	public static of(project: Project): Vitest | undefined {
		const isVitest = (c: Component): c is Vitest => c instanceof Vitest
		return project.components.find(isVitest)
	}

	readonly configType: VitestConfigType
	readonly configFile: TypeScriptSourceConfig<UserWorkspaceConfig>
	readonly options: Required<VitestOptions>
	readonly vite?: Vite
	readonly testWatchTask: Task
	readonly #workspaceProjects: Map<string, Vitest> = new Map<string, Vitest>()

	constructor(
		public readonly project: TypeScriptProject,
		options: VitestOptions = {
			configFilePath: 'vitest.config.ts',
			configType: VitestConfigType.PROJECT,
			settings: {},
		}
	) {
		super(project)
		this.vite = Vite.of(this.project)
		this.options = deepMerge([
			{
				settings: {
					test: {
						name: project.name,
						include: [`${project.testdir}/\*\*/\*.spec.ts`],
					},
				},
			} as VitestOptions,
			options,
		]) as Required<VitestOptions>
		this.project.addDevDeps('vitest')
		this.configType = this.options.configType ?? VitestConfigType.PROJECT
		const configFilePath = this.options.configFilePath ?? 'vitest.config.ts'

		if (this.configType !== VitestConfigType.WORKSPACE) {
			this.project.testTask.exec('vitest', {
				args: ['--run'],
				receiveArgs: true,
			})
		}
		this.testWatchTask =
			this.project.tasks.tryFind('test:watch') ??
			this.project.addTask('test:watch', {
				receiveArgs: true,
				exec: 'vitest',
				description: 'Run tests on changes.',
			})

		this.project.tsconfigDev?.addInclude?.(configFilePath)
		this.project?.eslint?.addOverride?.({
			files: [
				configFilePath,
				...(this.options.settings?.test?.include ?? [this.project.testdir]),
			],
			rules: {
				'import/no-extraneous-dependencies': ['off'],
			},
		})

		this.configFile =
			TypeScriptSourceConfig.withCallExpressionConfig<UserWorkspaceConfig>(
				this.project,
				configFilePath,
				{
					source: `export default ${this.defineName}({})`,
					recreate: true,
					marker: true,
					config: this.options.settings,
				}
			)

		this.configFile.addImport({
			moduleSpecifier: 'vitest/config',
			namedImports: [this.defineName],
		})

		if (this.configType === VitestConfigType.PROJECT) {
			this.tryFindWorkspace()?.addProjectConfig?.(this)
		}

		if (this.configType === VitestConfigType.WORKSPACE) {
			new JsonFile(this.project, 'vitest.workspace.json', {
				readonly: true,
				allowComments: false,
				obj: () => Array.from(this.#workspaceProjects.keys()),
			})
		}
	}

	/**
	 * Get the define function name.
	 * @protected
	 */
	protected get defineName(): string {
		return this.configType === VitestConfigType.PROJECT
			? 'defineProject'
			: 'defineConfig'
	}

	/**
	 * Merge vite config if applicable.
	 * @protected
	 */
	protected mergeVite(): void {
		if (!this.vite) return
		this.configFile.addImport({
			moduleSpecifier: 'vitest/config',
			namedImports: ['mergeConfig'],
		})
		this.configFile.addImport({
			moduleSpecifier: cwdRelativePath(
				this.project.outdir,
				this.vite.file.absolutePath.split('.').slice(0, -1).join('.')
			),
			defaultImport: 'viteConfig',
		})
		this.configFile.addConfigTransform((cfg, src) => {
			src.addExportAssignment({
				expression: (writer) =>
					writer.write(
						`mergeConfig(viteConfig, ${this.defineName}(${cfg.getFullText()}))`
					),
				isExportEquals: false,
			})
			src.removeDefaultExport()
		})
	}

	/**
	 * Try to find the workspace {@link Vitest} component.
	 */
	tryFindWorkspace(): Vitest | undefined {
		if (this.configType === VitestConfigType.WORKSPACE) return this
		if (this.project.parent) {
			return Vitest.of(this.project.parent)?.tryFindWorkspace?.()
		}
		return
	}

	/**
	 * Add {@link Vitest} component to this workspace.
	 * @param component Target component.
	 */
	addProjectConfig(component: Vitest): this {
		if (this.configType !== VitestConfigType.WORKSPACE)
			throw new Error('No vitest workspace found. Cannot add project config!')
		const relPath = path.format({
			dir: '.',
			base: path.relative(
				path.dirname(this.configFile.absolutePath),
				component.configFile.absolutePath
			),
		})
		this.#workspaceProjects.set(relPath, component)
		return this
	}

	/**
	 * Add config to merge.
	 * @param config Config part.
	 */
	addConfig(config: this['configFile']['__mergeSchema']): this {
		this.configFile.addConfig(config)
		return this
	}

	/**
	 * Add plugin to merge.
	 * @param plugin Plugin part.
	 */
	addPlugin(plugin: this['configFile']['__pluginSchema']) {
		this.configFile.addPlugin(plugin)
		return this
	}

	/**
	 * @inheritDoc
	 */
	preSynthesize() {
		super.preSynthesize()
		this.mergeVite()
	}
}
