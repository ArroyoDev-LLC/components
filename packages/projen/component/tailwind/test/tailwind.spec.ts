import os from 'node:os'
import path from 'node:path'
import { typescript } from 'projen'
import { Testing } from 'projen/lib/testing'
import { beforeEach, expect, test } from 'vitest'
import { Tailwind } from '../src'

interface TestContext {
	project: typescript.TypeScriptProject
}

beforeEach<TestContext>((ctx) => {
	ctx.project = new typescript.TypeScriptProject({
		name: ctx.meta.id,
		outdir: path.join(os.tmpdir(), ctx.meta.id),
		defaultReleaseBranch: 'main',
	})
})

test<TestContext>('renders as expected.', async (ctx) => {
	const tailwind = new Tailwind(ctx.project)
	const synth = Testing.synth(ctx.project)
	expect(synth['tailwind.config.mjs']).toBeDefined()
	expect(synth['tailwind.config.mjs']).toMatchSnapshot()
	expect(synth['tsconfig.dev.json']).toMatchSnapshot()
	expect(synth['package.json']).toMatchSnapshot()
})

test<TestContext>('renders with plugins as expected.', async (ctx) => {
	const tailwind = new Tailwind(ctx.project)
	tailwind.addPlugin({
		name: 'forms',
		moduleImport: {
			moduleSpecifier: '@tailwindcss/forms',
			defaultImport: 'forms',
		},
	})
	const synth = Testing.synth(ctx.project)
	expect(synth['tailwind.config.js']).toMatchSnapshot()
	expect(synth['package.json']).toHaveProperty([
		'dependencies',
		'@tailwindcss/forms',
	])
})

test<TestContext>('merges options as expected.', async (ctx) => {
	const tailwind = new Tailwind(ctx.project, {
		filePath: 'tailwind.config.js',
		config: {
			content: ['my', 'files'],
			prune: ['erethang'],
			important: true,
			theme: {
				extend: {
					colors: {
						primary: 'blue',
					},
				},
			},
		},
	})
	expect(tailwind.options).toMatchInlineSnapshot(`
		{
		  "config": {
		    "content": [
		      "my",
		      "files",
		    ],
		    "important": true,
		    "prune": [
		      "erethang",
		    ],
		    "theme": {
		      "extend": {
		        "colors": {
		          "primary": "blue",
		        },
		      },
		    },
		  },
		  "filePath": "tailwind.config.js",
		}
	`)
})
