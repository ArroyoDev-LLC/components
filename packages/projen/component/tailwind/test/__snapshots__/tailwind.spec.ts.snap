// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders as expected. 1`] = `
"// ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

/** @type {import('tailwindcss').Config} */
export default {
    content: [\\"./index.html\\",  \\"./src/**/*.{vue,ts,tsx}\\"],
    theme: {
        extend: {}
    }
}"
`;

exports[`renders as expected. 2`] = `
{
  "compilerOptions": {
    "alwaysStrict": true,
    "declaration": true,
    "esModuleInterop": true,
    "experimentalDecorators": true,
    "inlineSourceMap": true,
    "inlineSources": true,
    "lib": [
      "es2019",
    ],
    "module": "CommonJS",
    "noEmitOnError": false,
    "noFallthroughCasesInSwitch": true,
    "noImplicitAny": true,
    "noImplicitReturns": true,
    "noImplicitThis": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "resolveJsonModule": true,
    "strict": true,
    "strictNullChecks": true,
    "strictPropertyInitialization": true,
    "stripInternal": true,
    "target": "ES2019",
  },
  "exclude": [
    "node_modules",
  ],
  "include": [
    ".projenrc.js",
    "src/**/*.ts",
    "test/**/*.ts",
    "tailwind.config.mjs",
  ],
}
`;

exports[`renders as expected. 3`] = `
{
  "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
  "dependencies": {
    "tailwindcss": "*",
  },
  "devDependencies": {
    "@types/jest": "*",
    "@types/node": "^16",
    "@typescript-eslint/eslint-plugin": "^6",
    "@typescript-eslint/parser": "^6",
    "eslint": "^8",
    "eslint-import-resolver-node": "*",
    "eslint-import-resolver-typescript": "*",
    "eslint-plugin-import": "*",
    "jest": "*",
    "jest-junit": "^15",
    "npm-check-updates": "^16",
    "projen": "*",
    "standard-version": "^9",
    "ts-jest": "*",
    "typescript": "*",
  },
  "jest": {
    "clearMocks": true,
    "collectCoverage": true,
    "coverageDirectory": "coverage",
    "coveragePathIgnorePatterns": [
      "/node_modules/",
    ],
    "coverageReporters": [
      "json",
      "lcov",
      "clover",
      "cobertura",
      "text",
    ],
    "globals": {
      "ts-jest": {
        "tsconfig": "tsconfig.dev.json",
      },
    },
    "preset": "ts-jest",
    "reporters": [
      "default",
      [
        "jest-junit",
        {
          "outputDirectory": "test-reports",
        },
      ],
    ],
    "testMatch": [
      "<rootDir>/src/**/__tests__/**/*.ts?(x)",
      "<rootDir>/(test|src)/**/*(*.)@(spec|test).ts?(x)",
    ],
    "testPathIgnorePatterns": [
      "/node_modules/",
    ],
    "watchPathIgnorePatterns": [
      "/node_modules/",
    ],
  },
  "license": "Apache-2.0",
  "main": "lib/index.js",
  "name": "-1924646760_0",
  "scripts": {
    "build": "npx projen build",
    "bump": "npx projen bump",
    "clobber": "npx projen clobber",
    "compile": "npx projen compile",
    "default": "npx projen default",
    "eject": "npx projen eject",
    "eslint": "npx projen eslint",
    "package": "npx projen package",
    "post-compile": "npx projen post-compile",
    "post-upgrade": "npx projen post-upgrade",
    "pre-compile": "npx projen pre-compile",
    "projen": "npx projen",
    "release": "npx projen release",
    "test": "npx projen test",
    "test:watch": "npx projen test:watch",
    "unbump": "npx projen unbump",
    "upgrade": "npx projen upgrade",
    "watch": "npx projen watch",
  },
  "types": "lib/index.d.ts",
  "version": "0.0.0",
}
`;

exports[`renders with plugins as expected. 1`] = `undefined`;
